QPS: Reduce number of arguments of convergence tests.

haplav opened the pull request:
Solved QP, iteration and residual norm are available from the QPS struct.

haplav committed:
QPS: Reduce number of arguments of convergence tests.

Solved QP, iteration and residual norm are available from the QPS struct.

jkruzik commented on src/qps/impls/smalxe/smalxe.c:
```diff
@@ -616,19 +616,22 @@ PETSC_STATIC_INLINE PetscErrorCode QPSConverged_Inner_SMALXE_Monitor_Inner(QPS q
+  FLLOP_ASSERT(qp_inner->parent == cctx->qp_outer,"qp_inner->parent == qp_outer");
+  FLLOP_ASSERT(qp_inner->x == cctx->qp_outer->x, "qp_inner->x == qp_outer->x");
   TRY( PetscObjectGetComm((PetscObject)qps_inner,&comm) );
+  if (!qps_inner->setupcalled) FLLOP_SETERRQ(comm, PETSC_ERR_ARG_WRONGSTATE, "Inner QPSSetUp() not yet called");
```
I do not think that this is neccessary.

haplav replied:
but does not hurt neither - kind of assertion check

jkruzik replied:
My point is that we know that QPSSetUp(qps_inner) was called because at this point QPSSetUp(qps_outter) was called and created qps_inner.

haplav replied:
ok, I moved all the checks here to Solve and only in debug mode

jkruzik commented on src/qps/interface/qps.c:
```diff
@@ -683,14 +681,19 @@ PetscErrorCode QPSGetConvergenceContext(QPS qps,void **ctx)
 
   PetscFunctionBegin;
   PetscValidHeaderSpecific(qps,QPS_CLASSID,1);
+  PetscValidHeaderSpecific(qp,QP_CLASSID,2);
```
```suggestion
```

haplav replied:
oops

jkruzik commented on src/qps/interface/qps.c:
```diff
@@ -683,14 +681,19 @@ PetscErrorCode QPSGetConvergenceContext(QPS qps,void **ctx)
+  PetscValidHeaderSpecific(qp,QP_CLASSID,2);
   *reason = KSP_CONVERGED_ITERATING;
 
+  if (!qps->setupcalled) FLLOP_SETERRQ(((PetscObject) qps)->comm, PETSC_ERR_ARG_WRONGSTATE, "QPSSetUp() not yet called");
```
Again, is this necessary?

haplav replied:
Here it's even more important. We need to be sure e.g. that `solQP` is in place.

jkruzik replied:
OK. However, I would suggest putting this in QPSConvergedDefaultSetUp(), which also computes norm of solQP->b and then this check will be done only once.

haplav replied:
OK, this makes sense.

haplav committed:
Update src/qps/interface/qps.c

Co-authored-by: Jakub Kruzik <jakub.kruzik@gmail.com>

jkruzik commented on src/qps/interface/qps.c:
```diff
@@ -683,14 +681,19 @@ PetscErrorCode QPSGetConvergenceContext(QPS qps,void **ctx)
+  if (!qps->setupcalled) FLLOP_SETERRQ(((PetscObject) qps)->comm, PETSC_ERR_ARG_WRONGSTATE, "QPSSetUp() not yet called");
   if (!cctx->setup_called) {
     TRY( QPSConvergedDefaultSetUp(ctx,qps) );
   }
```
I can't comment on the line bellow, but this `if` should be swapped with the following one.

haplav replied:
yes

jkruzik commented:
Do we need `ctx` as an argument to the convergence test? It is stored in `qps->cnvctx`, and we have set/get functions for it.

haplav commented:
> Do we need `ctx` as an argument to the convergence test? It is stored in `qps->cnvctx`, and we have set/get functions for it.

True, let's remove this arg as well.

The only advantage of this arg might be that no call to the context getter is needed but source files which implement stopping criterions will usually include private headers so `qps->cnvctx` is directly accessible anyway.

haplav committed:
squash! QPS: Reduce number of arguments of convergence tests.

QPSSMALXE: move checks from QPSConverged_Inner_SMALXE to QPSSolve_SMALXE, do them only in debug mode
QPSConvergedDefault: fix check order

haplav commented:
> Do we need ctx as an argument to the convergence test? It is stored in qps->cnvctx, and we have set/get functions for it.

meta-note: better start off a thread next time in such cases

haplav committed:
squash! QPS: Reduce number of arguments of convergence tests.

Get rid of ctx arg as well.

haplav commented:
@jkruzik anything else? If it's good to go already, I prefer to squash it on my own to decide about the commit message.

jkruzik reviewed (approved):
Looks good!

jkruzik commented:
> > Do we need ctx as an argument to the convergence test? It is stored in qps->cnvctx, and we have set/get functions for it.
> 
> meta-note: better start off a thread next time in such cases

I do not know what you mean. There wasn't any thread related to this.

haplav commented on src/qps/impls/smalxe/smalxe.c:
```diff
@@ -616,18 +616,18 @@ PETSC_STATIC_INLINE PetscErrorCode QPSConverged_Inner_SMALXE_Monitor_Inner(QPS q
 #undef __FUNCT__
 #define __FUNCT__ "QPSConverged_Inner_SMALXE"
-PetscErrorCode QPSConverged_Inner_SMALXE(QPS qps_inner,QP qp_inner,PetscInt i,PetscReal gnorm,KSPConvergedReason *reason,void *ctx)
+PetscErrorCode QPSConverged_Inner_SMALXE(QPS qps_inner,KSPConvergedReason *reason)
```
I meant you could start a new thread e.g. here. I like threads more if applicable because they can be marked as resolved and collapsed.

haplav replied:
@jkruzik
